app/wasm/test/messages_test.go

```
@@ -336,3 +336,170 @@ func TestSwap(t *testing.T) {
+	uatom2 := poolFunds2[0].Amount.ToDec().MustFloat64()
+	uosmo2 := poolFunds2[1].Amount.ToDec().MustFloat64()
+	swapRate2 := uosmo2 / uatom2
+
+	amount := amountIn.Input.ToDec().MustFloat64()
+	slippage := 0.992 // FIXME: Derive this value from the pools and traded amounts
+	atomAmount := sdk.NewInt(int64(amount / swapRate1 / swapRate2 * slippage))
+
+	atomSwapAmount := wasmbindings.SwapAmount{Out: &atomAmount}
+
+	specs := map[string]struct {
+		swap    *wasmbindings.SwapMsg
+		expCost *wasmbindings.SwapAmount
+		expErr  bool
+	}{
+		"valid swap (exact in, 2 step multi-hop)": {
+			swap: &wasmbindings.SwapMsg{
+				First: wasmbindings.Swap{
+					PoolId:   starPool,
+					DenomIn:  "ustar",
+					DenomOut: "uosmo",
+				},
+				Route: []wasmbindings.Step{{
+					PoolId:   atomPool,
+					DenomOut: "uatom",
+				}},
+				Amount: wasmbindings.SwapAmountWithLimit{
+					ExactIn: &amountIn,
+				},
+			},
+			expCost: &atomSwapAmount,
+		},
+		"non-existent step pool id": {
+			swap: &wasmbindings.SwapMsg{
+				First: wasmbindings.Swap{
+					PoolId:   starPool,
+					DenomIn:  "ustar",
+					DenomOut: "uosmo",
+				},
+				Route: []wasmbindings.Step{{
+					PoolId:   atomPool + 2,
+					DenomOut: "uatom",
+				}},
+				Amount: wasmbindings.SwapAmountWithLimit{
+					ExactIn: &amountIn,
+				},
+			},
+			expErr: true,
+		},
+		"zero step pool id": {
+			swap: &wasmbindings.SwapMsg{
+				First: wasmbindings.Swap{
+					PoolId:   starPool,
+					DenomIn:  "ustar",
+					DenomOut: "uosmo",
+				},
+				Route: []wasmbindings.Step{{
+					PoolId:   0,
+					DenomOut: "uatom",
+				}},
+				Amount: wasmbindings.SwapAmountWithLimit{
+					ExactIn: &amountIn,
+				},
+			},
+			expErr: true,
+		},
+		"wrong step denom out": {
+			swap: &wasmbindings.SwapMsg{
+				First: wasmbindings.Swap{
+					PoolId:   starPool,
+					DenomIn:  "ustar",
+					DenomOut: "uosmo",
+				},
+				Route: []wasmbindings.Step{{
+					PoolId:   atomPool,
+					DenomOut: "ATOM",
```

It would be interesting to test "uosmo" here, where we would be requesting a uosmo -> uosmo swap.
I assume this should return an error in gamm, but good to test